[ty] Add diagnostics for invalid `await` expressions

theammir opened the pull request:
<!--
Thank you for contributing to Ruff/ty! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title? (Please prefix with `[ty]` for ty pull
  requests.)
- Does this pull request include references to any relevant issues?
-->

## Summary

<!-- What's the purpose of the change? What does it do, and why? -->
This PR adds a new lint, `invalid-await`, for all sorts of reasons why an object may not be `await`able, as discussed in astral-sh/ty#919.
Precisely, `__await__` is guarded against being missing, possibly unbound, or improperly defined (expects additional arguments or doesn't return an iterator).

Of course, diagnostics need to be fine-tuned.  If `__await__` cannot be called with no extra arguments, it indicates an error (or a quirk?) in the method signature, not at the call site. Without any doubt, such an object is not `Awaitable`, but I feel like talking about arguments for an *implicit* call is a bit leaky.
I didn't reference any actual diagnostic messages in the lint definition, because I want to hear feedback first.

Also, there's no mention of the actual required method signature for `__await__` anywhere in the docs. The only reference I had is the `typing` stub. I basically ended up linking `[Awaitable]` to ["must implement `__await__`"](https://docs.python.org/3/library/collections.abc.html#collections.abc.Awaitable), which is insufficient on its own.

## Test Plan

<!-- How was it tested? -->
The following code was tested:
```python
import asyncio
import typing


class Awaitable:
    def __await__(self) -> typing.Generator[typing.Any, None, int]:
        yield None
        return 5


class NoDunderMethod:
    pass


class InvalidAwaitArgs:
    def __await__(self, value: int) -> int:
        return value


class InvalidAwaitReturn:
    def __await__(self) -> int:
        return 5


class InvalidAwaitReturnImplicit:
    def __await__(self):
        pass


async def main() -> None:
    result = await Awaitable()  # valid
    result = await NoDunderMethod()  # `__await__` is missing
    result = await InvalidAwaitReturn()  # `__await__` returns `int`, which is not a valid iterator 
    result = await InvalidAwaitArgs()  # `__await__` expects additional arguments and cannot be called implicitly
    result = await InvalidAwaitReturnImplicit()  # `__await__` returns `Unknown`, which is not a valid iterator


asyncio.run(main())
```

theammir commented:
Alright, now
```python
class InvalidAwaitReturnImplicit:
    def __await__(self):
        pass
```
is awaitable as far as the type signature goes.

Also, union types don't work. I suppose I have to check every element for being a generator, and construct a union of possible return types if all of them are, in fact, awaitable.

sharkdp commented:
> Alright, now
> 
> ```python
> class InvalidAwaitReturnImplicit:
>     def __await__(self):
>         pass
> ```
> 
> is awaitable as far as the type signature goes.

:+1: 

> Also, union types don't work. I suppose I have to check every element for being a generator, and construct a union of possible return types if all of them are, in fact, awaitable.

Ah, yes, you're right. Thank you for fixing these. I think you should be able to use `UnionType::map` to recursively apply `generator_return_type` to all elements of the union.

theammir commented:
Now distinguishes between
```python
class AwaitableUnion:  # Generator | Unknown, awaitable
    if datetime.today().weekday() == 6:

        def __await__(self) -> typing.Generator[typing.Any, None, None]:
            yield

    else:

        def __await__(self):
            pass


class UnawaitableUnion:  # Generator | int, non-awaitable
    if datetime.today().weekday() == 6:

        def __await__(self) -> typing.Generator[typing.Any, None, None]:
            yield

    else:

        def __await__(self) -> int:
            return 5
```

carljm commented on crates/ty_python_semantic/src/types.rs:
```diff
@@ -6802,6 +6813,51 @@ impl<'db> TypeVarBoundOrConstraints<'db> {
+        );
+        match self {
+            Self::Call(CallDunderError::CallError(..)) => {
+                // TODO: Indicate definition error, not usage error.
```
I think it's reasonable to emit a diagnostic at the callsite here. It might also be nice to emit a diagnostic where `__await__` is defined with an incorrect signature, but I would see this as in addition, not instead of.

carljm commented on crates/ty_python_semantic/src/types.rs:
```diff
@@ -6802,6 +6813,51 @@ impl<'db> TypeVarBoundOrConstraints<'db> {
+        let mut diag = builder.into_diagnostic(
+            format_args!("`{type}` is not awaitable", type = context_expression_type.display(db)),
+        );
+        match self {
+            Self::Call(CallDunderError::CallError(..)) => {
+                // TODO: Indicate definition error, not usage error.
+                diag.info(
+                    "`__await__` expects additional arguments and cannot be called implicitly",
+                );
+            }
+            Self::Call(CallDunderError::PossiblyUnbound(..)) => {
+                diag.info("`__await__` is possibly unbound");
+            }
+            Self::Call(CallDunderError::MethodNotAvailable) => diag.info("`__await__` is missing"),
+            Self::InvalidReturnType(return_type) => diag.info(format_args!(
+                "`__await__` returns `{return_type}`, which is not a valid iterator",
+                return_type = return_type.display(db)
+            )),
+        }
+    }
```
Is it possible that in these cases we can highlight the range of the `__await__` method, or of the definition of the type that is lacking an `__await__` method?

carljm commented on crates/ty_python_semantic/src/types/diagnostic.rs:
```diff
@@ -552,6 +553,36 @@ declare_lint! {
+    ///         return 5
+    ///
+    /// async def main() -> None:
+    ///     await InvalidAwait()  # error: TODO:
+    ///     await 42  # error: TODO:
```
why are there TODOs here?

theammir replied:
Wanted to hear feedback before documenting actual error messages, I think.

carljm reviewed (commented):
This is looking pretty good, thank you! Can we add some mdtests (maybe in a new `diagnostics/invalid_await.md`) demonstrating cases where this diagnostic is emitted, and snapshotting their output?

theammir commented:
Added some useful secondary annotations for the lint. Now `report_diagnostic` is a huge match-case, but from what I've seen, a lot of `ty` is, too.

Added mdtests as proposed. Some other tests that rely on `await` expressions have been broken, for instance:
https://github.com/astral-sh/ruff/blob/3a542a80f65d66a15e22b6fe161e5af05ddffa83/crates/ty_python_semantic/resources/mdtest/diagnostics/semantic_syntax_errors.md?plain=1#L136-L138
Would expecting `error: [invalid-await]` here as well break atomicity?

Even though providing the type definition of a non-awaitable object is useful, sometimes diagnostics peek into built-ins:
```
error[invalid-await]: `Literal[1]` is not awaitable
   --> test-project/main.py:83:11
    |
 81 |     await Awaitable()  # valid
 82 |     await AwaitableUnion()  # valid
 83 |     await 1  # invalid
    |           ^
 84 |
 85 |     await NonCallableAwait()  # invalid
    |
   ::: stdlib/builtins.pyi:337:7
    |
335 | _LiteralInteger = _PositiveInteger | _NegativeInteger | Literal[0]  # noqa: Y026  # TODO: Use TypeAlias once mypy bugs are fixed
336 |
337 | class int:
    |       --- type defined here
338 |     """int([x]) -> integer
339 |     int(x, base=10) -> integer
    |
info: `__await__` is missing
info: rule `invalid-await` is enabled by default
```
---
Also, while fooling around, I discovered that awaiting an instance of this type is perfectly valid, and it isn't considered possibly unbound:
```py
class PossiblyUnbound:
    if datetime.today().weekday() == 0:

        def __await__(self):
            yield 1

    elif datetime.today().weekday() == 1:

        def __await__(self):
            yield 2
```
I wonder why.

carljm commented:
> Also, while fooling around, I discovered that awaiting an instance of this type is perfectly valid, and it isn't considered possibly unbound

Thanks for this comment! I explored it and found a bug, fixed in https://github.com/astral-sh/ruff/pull/19884

carljm commented:
@theammir Just for future reference (in case you make future PRs), you may want to note some things I'm fixing in this PR:

* `cargo dev generate-all` (because this PR adds a new diagnostic and we need to generate some rules metadata)
* Adding `<!-- snapshot-diagnostics -->` to the mdtest so we get full diagnostic snapshots, then running `cargo insta test -p ty_python_semantic` followed by `cargo insta review` to approve those snapshots.
* Allowing `pre-commit` to auto-format the mdtests.

theammir commented:
> It's not clear to me that using the protocol is a better approach to get good diagnostics here, but it's something we could consider in future. For now I think this PR is good.

My entire reasoning behind the design I went with was trying to build from the blocks available, and that one todo comment in `__enter__`-related functionality that says that we should check for assignability once we support protocols in the first place.

That said, pyright's "is not awaitable" diagnostics are thorough, but also deeply nested and somewhat convoluted:
```python
Pyright: "InvalidAwaitArgs" is not awaitable
     "InvalidAwaitArgs" is incompatible with protocol "Awaitable[_T_co@Awaitable]"
       "__await__" is an incompatible type
         Type "(value: int) -> Generator[int, Any, None]" is not assignable to type "() -> Generator[Any, Any, _T_co@Awaitable]"
           Extra parameter "value" [reportGeneralTypeIssues]
```
All the info about specific extra parameters is there (if you are careful enough to find it on line 4, otherwise read even further), and there's more to the protocol error message than just reporting unassignability anyway.
Just thoughts.

> Just for future reference (in case you make future PRs), you may want to note some things I'm fixing in this PR:

Thanks! It's been really fun so far, I'd stick around. :)
